bgpd/bgp_linkstate_tlv.c

```diff
@@ -0,0 +1,1090 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/* BGP Link-State TLV Serializer/Deserializer
+ * Copyright 2023 6WIND S.A.
+ */
+
+#include <zebra.h>
+
+#include "iso.h"
+
+#include "bgpd/bgpd.h"
+#include "bgpd/bgp_route.h"
+#include "bgpd/bgp_debug.h"
+#include "bgpd/bgp_errors.h"
+#include "bgpd/bgp_linkstate.h"
+#include "bgpd/bgp_linkstate_tlv.h"
+
+DEFINE_MTYPE_STATIC(BGPD, BGP_LS_TYPE_NODE, "BGP-LS Type Node");
+DEFINE_MTYPE_STATIC(BGPD, BGP_LS_TYPE_LINK, "BGP-LS Type Link");
+DEFINE_MTYPE_STATIC(BGPD, BGP_LS_TYPE_PREFIX4, "BGP-LS Type Prefix IPv4");
+DEFINE_MTYPE_STATIC(BGPD, BGP_LS_TYPE_PREFIX6, "BGP-LS Type Prefix IPv6");
+
+struct bgp_linkstate_tlv_info {
+	const char *descr;
+	uint8_t min_size;
+	uint16_t max_size;
+	uint8_t multiple;
+};
+
+#define MAX_SZ 0xFFFF
+#define UNDEF_MULTPL 1
+
+struct bgp_linkstate_tlv_info bgp_linkstate_tlv_infos[BGP_LS_TLV_MAX] = {
+		/* NLRI TLV */
+		[BGP_LS_TLV_LOCAL_NODE_DESCRIPTORS] = {"Local Node Descriptors", 1, MAX_SZ, UNDEF_MULTPL},
+		[BGP_LS_TLV_REMOTE_NODE_DESCRIPTORS] = {"Remote Node Descriptors", 1, MAX_SZ, UNDEF_MULTPL},
+		[BGP_LS_TLV_LINK_LOCAL_REMOTE_IDENTIFIERS] = {"Link Local/Remote Identifiers", 2, 2, UNDEF_MULTPL},
+		[BGP_LS_TLV_IPV4_INTERFACE_ADDRESS] = {"IPv4 interface address", 4, 4, UNDEF_MULTPL},
+		[BGP_LS_TLV_IPV4_NEIGHBOR_ADDRESS] = {"IPv4 neighbor address", 4, 4, UNDEF_MULTPL},
+		[BGP_LS_TLV_IPV6_INTERFACE_ADDRESS] = {"IPv6 interface address", 16, 16, UNDEF_MULTPL},
+		[BGP_LS_TLV_IPV6_NEIGHBOR_ADDRESS] = {"IPv6 neighbor address", 16, 16, UNDEF_MULTPL},
+		[BGP_LS_TLV_OSPF_ROUTE_TYPE] = {"OSPF Route Type", 1, 1, UNDEF_MULTPL},
+		[BGP_LS_TLV_IP_REACHABILITY_INFORMATION] = {"IP Reachability Information", 2, 17, UNDEF_MULTPL},
+		[BGP_LS_TLV_AUTONOMOUS_SYSTEM] = {"Autonomous System", 4, 4, UNDEF_MULTPL},
+		[BGP_LS_TLV_BGP_LS_IDENTIFIER] = {"BGP-LS Identifier", 4, 4, UNDEF_MULTPL},
+		[BGP_LS_TLV_OSPF_AREA_ID] = {"OSPF Area-ID", 4, 4, UNDEF_MULTPL},
+		[BGP_LS_TLV_IGP_ROUTER_ID] = {"IGP Router-ID", 4, 8, UNDEF_MULTPL},
+		/* NRLI & BGP-LS Attributes */
+		[BGP_LS_TLV_MULTI_TOPOLOGY_ID] = {"Multi-Topology ID", 2, MAX_SZ, 2},
+};
+
+/* Return the TLV length is valid for the TLV type */
+static bool bgp_ls_tlv_check_size(enum bgp_linkstate_tlv type, size_t length)
+{
+	if (type > BGP_LS_TLV_MAX ||
+	    bgp_linkstate_tlv_infos[type].descr == NULL)
+		/* TLV type is not defined. Cannot check size */
+		return false;
+
```

Comment:
return true, assuming that the length param is ok with the new type.

Reply:
see https://github.com/FRRouting/frr/pull/12649#discussion_r1169925019